onnxruntime/core/providers/cpu/tensor/upsample.h

```diff
@@ -378,7 +397,73 @@ class Upsample : public UpsampleBase, public OpKernel {
                      const gsl::span<const int64_t>& output_dims) const;
 };
 
+BilinearParams SetupUpsampleBilinear(const int64_t input_height,
+                                     const int64_t input_width,
+                                     const int64_t output_height,
+                                     const int64_t output_width,
+                                     const float height_scale,
+                                     const float width_scale,
+                                     const std::vector<float>& roi,
+                                     AllocatorPtr& alloc,
+                                     const GetOriginalCoordinateFunc& get_original_coordinate,
+                                     bool is_nchw);
+
+template <typename T>
+void NhwcUpsampleBilinear(const int64_t batch_size,
+                          const int64_t num_channels,
+                          const int64_t input_height,
+                          const int64_t input_width,
+                          const int64_t output_height,
+                          const int64_t output_width,
+                          const float height_scale,
+                          const float width_scale,
+                          const std::vector<float>& roi,
+                          const bool use_extrapolation,
+                          const float extrapolation_value,
+                          const T* const XdataBase,
+                          T* const YdataBase,
+                          AllocatorPtr& alloc,
+                          const GetOriginalCoordinateFunc& get_original_coordinate,
+                          concurrency::ThreadPool* tp) {
+  BilinearParams p = SetupUpsampleBilinear(input_height, input_width, output_height, output_width,
+                                           height_scale, width_scale, roi,
+                                           alloc, get_original_coordinate, false);
+  for (int64_t n = 0; n < batch_size; ++n) {
+    const T* Xdata = XdataBase + n * (input_height * input_width) * num_channels;
+    T* Ydata = YdataBase + n * (output_height * output_width) * num_channels;
+    concurrency::ThreadPool::TryParallelFor(
+        tp, output_height * output_width,
+        static_cast<double>(input_height * input_width * output_height * output_width * num_channels),
```

Comment:
>static_cast<double>(input_height * input_width * output_height * output_width * num_channels), [](http://example.com/codeflow?start=8&length=94)

This cost_per_unit looks not right. It should be an estimation of the cost(cpu cycles) of a unit of work. You are giving the total cost. Should it be factor*num_channels, where factor is a const.

Reply:
Change the cost to factor * num_channels.